io_scene_godot/converters/mesh.py

```diff
@@ -115,6 +116,78 @@ def export_object_link_material(escn_file, export_settings, mesh_object,
                 )
 
 
+def get_applicable_modifiers(obj, export_settings):
+    ignore_modifiers = []
+    if not export_settings['use_mesh_modifiers']:
+        return []
+    if "ARMATURE" in export_settings['object_types']:
+        ignore_modifiers.append(bpy.types.ArmatureModifier)
+    ignore_modifiers = tuple(ignore_modifiers)
+    return [m for m in obj.modifiers if not isinstance(m, ignore_modifiers)
+            and m.show_viewport]
+
+
+def record_modifier_config(obj):
+    modifier_config_cache = []
+    for mod in obj.modifiers:
+        modifier_config_cache.append(mod.show_viewport)
+    return modifier_config_cache
+
+def restore_modifier_config(obj, modifier_config_cache):
+    for i, mod in enumerate(obj.modifiers):
+        mod.show_viewport = modifier_config_cache[i]
+
+def has_shape_keys(object_data):
+    return (hasattr(object_data, "shape_keys") and
+            object_data.shape_keys is not None)
+
+def get_bl_rna_properties(obj):
+    """Property serialization helper"""
+    values = []
+    for k in obj.bl_rna.properties.keys()[1:]:
+        values.append(get_bl_prop_value(obj.id_data, getattr(obj, k)))
+    return tuple(values)
+
+NON_IMMUTABLE = (mathutils.Color, mathutils.Euler, mathutils.Matrix,
+                 mathutils.Quaternion, mathutils.Vector)
+
+def get_bl_prop_value(id_data, prop):
+    """Property serialization helper"""
+    if isinstance(prop, NON_IMMUTABLE):
+        return prop[:]
+    if isinstance(prop, bpy.types.bpy_prop_array):
+        return tuple(prop)
+    if isinstance(prop, bpy.types.bpy_prop_collection):
+        return tuple([get_bl_prop_value(id_data, item) for item in prop])
+    if isinstance(prop, bpy.types.bpy_struct):
+        # if bpy_struct relates to parent object (ie, a propertygroup or
+        # similar structs containing properties) then it's embedded data
+        # we want to traverse, not a pointer to another object.
+        # NOTE: If I'm wrong about any of this, expect bugs!
+        if prop.id_data == id_data:
+            return get_bl_rna_properties(prop)
+    return prop
+
+class MeshResourceKey:
+    """Produces a key based on an object's data, its enabled modifiers
+    and their properties."""
+    # pylint: disable=protected-access
+    def __init__(self, obj, export_settings):
+        data = [obj.data]
+        for mod in get_applicable_modifiers(obj, export_settings):
+            mod_prop_values = get_bl_rna_properties(mod)
+            data.append(mod_prop_values)
+        # precalculate the hash now for better efficiency later
+        self._hash = hash(tuple(data))
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
```

Comment:
when comes to `__eq__`, I think it is usually not to compare the hash. Here you need to compare the original value (i.e. the tuple consists of objID and modifiers).  So that when different keys have a hash collision, we can still differentiate them with `__eq__`.

Reply:
I was only using __eq__ for testing, it's not used anywhere in the script (internally Dicts/hashtables will just use the __hash__ function, right?)
Anyway, considering it's a class with a single integer property it probably makes more sense to make MeshResourceKey a function for now.

Reply:
actually `__eq__` is necessary for a hash key. When two keys has the same hash value, their selves need to be compare in addition to the `__hash__`, because `__hash__` always has possibility (although very small) to collision.

Reply:
Ah ok. Didn't know that. Fixing it.